### PR TITLE
Convert `src/core/parser.js` to ES6 syntax and write more unit tests for the lexer and the parser

### PR DESCRIPTION
**Pull request by timvandermeij**
I recommend to review each commit with the `?w=1` flag to reduce the size of the diff.

### CODE REVIEW

File: src/core/parser.js [hunk 1]
```diff
@@ -1172,55 +1173,58 @@ class Lexer {
...
-      var hints = linDict.get('H'), hintsLength, item;
-      if (Array.isArray(hints) &&
-          ((hintsLength = hints.length) === 2 || hintsLength === 4)) {
-        for (var index = 0; index < hintsLength; index++) {
-          if (!(Number.isInteger(item = hints[index]) && item > 0)) {
-            throw new Error('Hint (' + index +
-                            ') in the linearization dictionary is invalid.');
+
+    function getHints(linDict) {
+      const hints = linDict.get('H'), hintsLength = hints.length;
```
Comment (Snuffleupagus) [inline src/core/parser.js]: Drive-by-comment: Please don't move `hintsLength = hints.length;` here, since that will result in errors being thrown if `hints` isn't actually an Array (which is why the old code checked that *first*).

File: test/unit/parser_spec.js [hunk 2]
```diff
@@ -14,110 +14,182 @@
...
-  describe('Lexer', function() {
-    it('should stop parsing numbers at the end of stream', function() {
-      const input = new StringStream('11.234');
-      const lexer = new Lexer(input);
-      expect(lexer.getNumber()).toEqual(11.234);
-    });
+  describe('Parser', function() {
+    describe('inlineStreamSkipEI', function() {
+      it('should skip over the EI marker if it is found', function() {
+        const string = 'BI /foo /bar EI<<';
```
Comment (Snuffleupagus) [inline test/unit/parser_spec.js]: While it may not matter for these particular tests, please note that these commands do *not* represent a valid content stream :-P

Unless you want to test for a *specific* mode of corruption, as found in real-world PDFs, it's always best if the unit-test is actually correct in my experience; please find an actual example below (from 10388_reduced.pdf). 
```
BI 
/W 16
/H 17
/BPC 1
/IM true 
ID ÿÿÿÿþþþþþþÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿ
EI
```

File: src/core/parser.js [hunk 3]
```diff
@@ -49,662 +50,705 @@ function computeAdler32(bytes) {
   return ((b % 65521) << 16) | (a % 65521);
 }
 
-var Parser = (function ParserClosure() {
-  function Parser(lexer, allowStreams, xref, recoveryMode) {
+class Parser {
+  constructor(lexer, allowStreams, xref, recoveryMode = false) {
```
Comment (Snuffleupagus) [inline src/core/parser.js]: Another possible improvement here, in a separate commit or PR, would be to change the constructor to take an object instead to get rid of the often "ugly" call-sites (and document them for free):
```javascript
constructor({ lexer, xref, allowStreams = false, recoveryMode = false, }) {
```

File: src/core/parser.js (hunk 1 above)
Comment (timvandermeij) [reply src/core/parser.js]: Whoops, good point! I'll adjust that.

File: test/unit/parser_spec.js (hunk 2 above)
Comment (timvandermeij) [reply test/unit/parser_spec.js]: That's a fair point. In this case it didn't really matter, but I agree that it's probably better to use real data instead of fake data here.

File: src/core/parser.js (hunk 3 above)
Comment (timvandermeij) [reply src/core/parser.js]: I agree, I'll see what I can do.

File: test/unit/parser_spec.js (hunk 2 above)
Comment (Snuffleupagus) [reply test/unit/parser_spec.js]: *Apparently I wrote that code, so here a bit of unsolicited advice :-)*

As the comment mentions this method is only used for inline image streams with an EOD (end-of-data) marker, i.e. `DCTDecode` (== JPEG), `ASCII85Decode`, and `ASCIIHexDecode`; see https://github.com/mozilla/pdf.js/blob/d587abbceb97008ed744261c219c90d3ce981f2b/src/core/parser.js#L390-L393
When this method is called, the EOD marker in question has been successfully found and we simply want to advance the stream position just after the `EI` command (ignoring any white-space in between).

The example above wasn't that great in hindsight, since in practice it wouldn't apply here (no /Filter entry, and thus no EOD marker present).
The following is unlikely to be actually valid image data, so please keep that in mind, but let's consider the following example instead:
```
q
1 0 0 1 0 0 cm 
BI 
/W 10
/H 10
/F /A85 
ID abc123~>
EI
Q
```
When the main part of `findASCII85DecodeInlineStreamEnd` has run, the stream position is now just after `~>`. With the help of `inlineStreamSkipEI` the stream position will be advanced to just after `EI`, which is what you need here.

File: src/core/parser.js (hunk 1 above)
Comment (timvandermeij) [reply src/core/parser.js]: Fixed in the new commit series.

File: test/unit/parser_spec.js (hunk 2 above)
Comment (timvandermeij) [reply test/unit/parser_spec.js]: Fixed in the new commit series. The unit test cases are now more like what we encounter in real streams. I couldn't find a nice way to put binary image data in there and I think the mock value `abc123` is fine since our only purpose is to show that we can skip over the `EI` marker, so the actual binary data seems irrelevant here.

File: src/core/parser.js (hunk 3 above)
Comment (timvandermeij) [reply src/core/parser.js]: Since this patch is already quite big, I'd like to leave this for a possible follow-up.

Comment (Snuffleupagus) [reply src/core/parser.js]: > Since this patch is already quite big, I'd like to leave this for a possible follow-up.

That's completely fine; just as long as it eventually happens, given how unwieldy the call-sites currently look :-)

**Review by Snuffleupagus — APPROVED** [review]
